### PR TITLE
Added DigraphOddGirth function

### PR DESCRIPTION
**Pull request by MTWhyte**
This pull request adds a function DigraphOddGirth, which calculates the odd girth of a digraph. The function may be later updated with an improved algorithm for digraphs with sparse adjacency matrices.

### CODE REVIEW

**Comment by MTWhyte** [comment]
This is the first time I've really independently used Git, so I apologise if I've made a mistake. If so, I'll ask for help and correct it.

File: gap/attr.gi [hunk 1]
```diff
@@ -954,6 +954,34 @@ function(digraph)
...
+function(digraph)
+  local A, B, k, N, girth, powerfound;
+  girth := DigraphGirth(digraph);
+  if girth = infinity or IsOddInt(girth) then
+    return girth;
+  fi;
+  N          := DigraphNrVertices(digraph);
+  A          := AdjacencyMatrix(digraph);
+  B          := Immutable(A ^ 2);
+  k          := 1;
```
Comment (wilfwilson) [inline gap/attr.gi]: If the girth is (say) `6`, then the odd girth is at least `7`, right? So why don't we skip checking whether the girth is `1`, `3`, or `5`?

**Comment by wilfwilson** [comment]
The code (including tests, and documentation) has to be formatted according to [gaplint](https://github.com/james-d-mitchell/gaplint) in order for the continuous integration to pass. Please could you tweak your code according to gaplint? Otherwise, and except for my question above, this PR looks good, thanks 🙂

File: gap/attr.gi [hunk 2]
```diff
@@ -954,6 +954,34 @@ function(digraph)
...
+  powerfound := false;
+  while k <= N + 2 do
+    A := A * B;
+    k := k + 2;
+    if Trace(A) <> 0 then
+      # It suffices to find the trace as entries of A are positive.
+      powerfound := true;
+      break;
```
Comment (james-d-mitchell) [inline gap/attr.gi]: The variable powerfound and this break aren’t necessary, you can just return k here, please amend!

File: gap/attr.gi [hunk 3]
```diff
@@ -954,6 +954,34 @@ function(digraph)
...
+[IsDigraph],
+function(digraph)
+  local A, B, k, N, girth, powerfound;
+  girth := DigraphGirth(digraph);
+  if girth = infinity or IsOddInt(girth) then
+    return girth;
+  fi;
+  N          := DigraphNrVertices(digraph);
+  A          := AdjacencyMatrix(digraph);
+  B          := Immutable(A ^ 2);
```
Comment (james-d-mitchell) [inline gap/attr.gi]: Is this immutable really necessary? I’m not sure

File: gap/attr.gi (hunk 1 above)
Comment (james-d-mitchell) [reply gap/attr.gi]: Good comment!

File: gap/attr.gi (hunk 2 above)
Comment (james-d-mitchell) [reply gap/attr.gi]: Also didn’t we discuss doing the powering thing by strongly connected component?

**Comment by MTWhyte** [comment]
Thanks for your comments - have implemented changes and will make a pull request in due course.

**Comment by james-d-mitchell** [comment]
The last thing is to add a test that shows that the by strongly connected component approach is actually faster. Such as:

~~~~
D := DigraphDisjointUnion(CycleDigraph(2), CycleDigraph(3));;
for i in [1 .. 1000] do
  D := DigraphDisjointUnion(D, CycleDigraph(3));
od;
OddGirth(D);
~~~~

**Comment by james-d-mitchell** [comment]
Looks like the CI is failing, but probably this is not your fault.

File: doc/attr.xml [hunk 4]
```diff
@@ -897,6 +897,43 @@ gap> DigraphGirth(gr);
...
 
+
+<#GAPDoc Label="DigraphOddGirth">
+<ManSection>
+  <Attr Name="DigraphOddGirth" Arg="digraph"/>
+  <Returns>An integer, or <K>infinity</K>.</Returns>
+  <Description>
+    This attribute returns the <E>odd girth</E> of the digraph <A>digraph</A>.
+    The <E>odd girth</E> of a digraph is the length of its shortest simple circuit of odd length.
+    See section <Ref Subsect="Definitions" Style="Number" /> for the definitions
```
Comment (james-d-mitchell) [inline doc/attr.xml]: "section" -> "Section"

File: doc/attr.xml [hunk 5]
```diff
@@ -897,6 +897,43 @@ gap> DigraphGirth(gr);
...
+3
+gap> gr := CycleDigraph(4);
+<digraph with 4 vertices, 4 edges>
+gap> DigraphOddGirth(gr);
+infinity
+gap> gr := Digraph([[2], [3], [], [3], [4]]);
+<digraph with 5 vertices, 4 edges>
+gap> DigraphOddGirth(gr);
+infinity
+
```
Comment (james-d-mitchell) [inline doc/attr.xml]: Remove this empty line please!

File: gap/attr.gi [hunk 6]
```diff
@@ -954,6 +954,50 @@ function(digraph)
...
 
+InstallMethod(DigraphOddGirth, "for a digraph",
+[IsDigraph],
+function(digraph)
+  local comps, girth, oddgirth, A, B, gr, k, comp;
+  if IsAcyclicDigraph(digraph) then
+    return infinity;
+  elif IsOddInt(DigraphGirth(digraph)) then
+    # No need to check girth isn't infinity, as we have
+    # that digraph is cyclic.
```
Comment (james-d-mitchell) [inline gap/attr.gi]: "cyclic" -> "not acyclic"

File: gap/attr.gi [hunk 7]
```diff
@@ -954,6 +954,50 @@ function(digraph)
...
+    return DigraphGirth(digraph);
+  fi;
+  comps := DigraphStronglyConnectedComponents(digraph).comps;
+  oddgirth := infinity;
+  for comp in comps do
+    if not IsStronglyConnectedDigraph(digraph) then
+      gr    := InducedSubdigraph(digraph, comp);
```
Comment (james-d-mitchell) [inline gap/attr.gi]: Weird spacing before `:=`

Comment (MTWhyte) [reply gap/attr.gi]: Thanks James, I have made all the changes and will push them now.

**Comment by james-d-mitchell** [comment]
Can you please rebase this onto the current `master` branch? @MTWhyte, thanks!

**Comment by MTWhyte** [comment]
@james-d-mitchell I think that I have successfully rebased it, but this is the first time I've attempted to do so independently, so I may be wrong. I'll try again if I've made a mistake.

**Comment by james-d-mitchell** [comment]
Thanks @MTWhyte, something doesn't look right though, only your commits should appear in the list of commits for the PR, but instead, there are 50 commits, many of which are already in the `master` branch. Can you perhaps try doing `git rebase master` again (it ought to not have any issues this time)?

**Pull request merged**